src/modules/dataman/dataman.cpp

```diff
@@ -53,113 +54,24 @@
-static int _file_initialize(unsigned max_offset);
-static void _file_shutdown();
-
-/* Private Ram based Operations */
-static ssize_t _ram_write(dm_item_t item, unsigned index, dm_persitence_t persistence, const void *buf,
-			  size_t count);
-static ssize_t _ram_read(dm_item_t item, unsigned index, void *buf, size_t count);
-static int  _ram_clear(dm_item_t item);
-static int  _ram_restart(dm_reset_reason reason);
-static int _ram_initialize(unsigned max_offset);
-static void _ram_shutdown();
+extern dm_operations_t dm_file_operations;
+extern dm_operations_t dm_ram_operations;
 
 #if defined(FLASH_BASED_DATAMAN)
-/* Private Ram_Flash based Operations */
-#define RAM_FLASH_FLUSH_TIMEOUT_USEC USEC_PER_SEC
-
-static ssize_t _ram_flash_write(dm_item_t item, unsigned index, dm_persitence_t persistence, const void *buf,
-				size_t count);
-static ssize_t _ram_flash_read(dm_item_t item, unsigned index, void *buf, size_t count);
-static int  _ram_flash_clear(dm_item_t item);
-static int  _ram_flash_restart(dm_reset_reason reason);
-static int _ram_flash_initialize(unsigned max_offset);
-static void _ram_flash_shutdown();
-static int _ram_flash_wait(px4_sem_t *sem);
+extern dm_operations_t dm_ram_flash_operations;
 #endif
 
-typedef struct dm_operations_t {
-	ssize_t (*write)(dm_item_t item, unsigned index, dm_persitence_t persistence, const void *buf, size_t count);
-	ssize_t (*read)(dm_item_t item, unsigned index, void *buf, size_t count);
-	int (*clear)(dm_item_t item);
-	int (*restart)(dm_reset_reason reason);
-	int (*initialize)(unsigned max_offset);
-	void (*shutdown)();
-	int (*wait)(px4_sem_t *sem);
-} dm_operations_t;
-
-static constexpr dm_operations_t dm_file_operations = {
-	.write   = _file_write,
-	.read    = _file_read,
-	.clear   = _file_clear,
-	.restart = _file_restart,
-	.initialize = _file_initialize,
-	.shutdown = _file_shutdown,
-	.wait = px4_sem_wait,
-};
+static dm_operations_t *g_dm_ops = nullptr;
 
-static constexpr dm_operations_t dm_ram_operations = {
-	.write   = _ram_write,
-	.read    = _ram_read,
-	.clear   = _ram_clear,
-	.restart = _ram_restart,
-	.initialize = _ram_initialize,
-	.shutdown = _ram_shutdown,
-	.wait = px4_sem_wait,
-};
-
-#if defined(FLASH_BASED_DATAMAN)
-static constexpr dm_operations_t dm_ram_flash_operations = {
-	.write   = _ram_flash_write,
-	.read    = _ram_flash_read,
-	.clear   = _ram_flash_clear,
-	.restart = _ram_flash_restart,
-	.initialize = _ram_flash_initialize,
-	.shutdown = _ram_flash_shutdown,
-	.wait = _ram_flash_wait,
-};
-#endif
-
-static const dm_operations_t *g_dm_ops;
-
-static struct {
-	union {
-		struct {
-			int fd;
-		} file;
-		struct {
-			uint8_t *data;
-			uint8_t *data_end;
-		} ram;
-#if defined(FLASH_BASED_DATAMAN)
-		struct {
-			uint8_t *data;
-			uint8_t *data_end;
-			/* sync above with RAM backend */
-			timespec flush_timeout;
-		} ram_flash;
-#endif
-	};
-	bool running;
-} dm_operations_data;
+dm_operations_data_t dm_operations_data{};
```

Comment:
The same here: either pass this as argument and make this `static`, or move the declaration into a common header.

Reply:
I've moved it to each type (file, ram, ram_flash) and split it to remove the unions.